lib/fastlane/plugin/wpmreleasetoolkit/actions/common/check_for_toolkit_updates_action.rb

```diff
@@ -0,0 +1,77 @@
+require 'fastlane/action'
+require 'rubygems/command_manager'
+
+module Fastlane
+  module Actions
+    class CheckForToolkitUpdatesAction < Action
+      TOOLKIT_SPEC_NAME = 'fastlane-plugin-wpmreleasetoolkit'.freeze
+
+      def self.run(params)
+        updater = Gem::CommandManager.instance[:update]
+        installed_gems = updater.highest_installed_gems.select { |spec| spec == TOOLKIT_SPEC_NAME }
+        local_version = Gem::Version.new(installed_gems[TOOLKIT_SPEC_NAME].version)
+        UI.message("Currently using release toolkit version #{local_version.to_s.yellow}.")
+        UI.message("Checking for updates now...")
+        updates_needed = updater.which_to_update(installed_gems, [TOOLKIT_SPEC_NAME])
+
+        if updates_needed.empty?
+          UI.success('Your release toolkit is up-to-date! ✅')
+          return nil
+        end
+
+        # Return type of which_to_update differs before/after RubyGems 3.1.0, so normalize to always use Gem::NameTuple
+        updates_needed = Gem::NameTuple.from_list(updates_needed)
+        latest_version = updates_needed.find { |gem_info| gem_info.name == TOOLKIT_SPEC_NAME }.version
+
+        UI.message(['There is a new version '.yellow, latest_version.to_s.red, ' of the release toolkit!'.yellow].join)
+        non_breaking_update = Gem::Requirement.new(local_version.approximate_recommendation).satisfied_by?(latest_version)
+        UI.important('The latest version introduces breaking changes!') unless non_breaking_update
```

Comment:
Note that even if the `latest_version` is incompatible with the `local_version` (breaking changes), we will still suggest to run `bundle update` – to at least get the latest non-breaking update aka latest version compatible with the `~> x.y` requirement we typically use in our `Pluginfile` and that is still safe to update to.

In those cases, after `bundle update` has run, we would still not be on the latest version (so running the check again will still warn that we're not on the latest), which is expected, because we don't want to accidentally pull breaking-changes in during that auto-update, but only the minor and patch changes. As intended by SemVer.

Instead, if the latest version is a breaking change, it's up to you to manually go modify your `Pluginfile` to update the `~> 1.0` to `~> 2.0`, then make the appropriate changes in your `Fastfile` according to the breaking changes described in the toolkit's CHANGELOG, then run `bundle install` and commit the major version bump.

Reply:
@AliSoftware I am a little confused by what the `latest_version` value will be if the installed version is `1.xx` and local version is `2.yy`. Will it be `1.aa` where `aa > xx`, so a non-breaking update, or will it be `2.yy`? Your comments suggest it'll be a minor bump, but I can't figure out why that's the case since [which_to_update seems to be finding the `highest_remote_version`](https://ruby-doc.org/stdlib-2.6.6/libdoc/rubygems/rdoc/Gem/Commands/UpdateCommand.html#method-i-which_to_update). (no idea what that actually means in this context)

I'd appreciate it if you can help me understand that. Thanks!

Reply:
No no you understood it correctly, `which_to_update` returns the highest version available regardless of semver, even if that's a major version.

So if we have currently `1.xx` installed but the latest version we released of the toolkit is 2.yy, `latest_version` is gonna be `2.yy`, not `1.aa`.

This is how I detect if there are breaking changes between the currently `local_version` installed version and the `latest_version`, because line 27 will see that `local_version` being `1.xx` is not compatible with `2.yy` (or more especially, that `local_version.approximate_recommendation`, aka `"~> 1.xx"` requirement, won't be satisfied by a `2.yy` version since that's not semver-compatible with `~>1.x`)

What might be confusing tho is the wording of the message maybe?

To clarify, here is what is supposed to be printed in the terminal if we are on `1.2.3` and the latest version available is `2.1.4`:
```
Currently using release toolkit version 1.2.3
Checking for updates now...
There is a new version 2.1.4 of the release toolkit!
The latest version introduces breaking changes!
Do you want to run bundle update now? y
$ bundle update fastlane-plugin-wpmreleasetoolkit
[…long log of bundle update there...]
fastlane-plugin-wpmreleasetoolkit have been updated. Please check and commit the changes in your Gemfile.lock file, then restart your previous invocation of fastlane to use the new toolkit.
```
BUT after those messages, the toolkit would have only been updated to something like `1.3.7` – the latest `1.x` version before the first `2.0.0`, because `bundle update` will respect the version constraint you put in your `Gemfile` (well, in the `Pluginfile` that is loaded from the `Gemfile`), which will supposedly be `'~> 1.0'` if we want to respect semver (using `~> x.y` in your `Gemfile` is the typical way to avoid accidentally bumping to a breaking version during `bundle update`, as it allows any version x.z where z>y but won't allow changing the x major number)

Reply:
I guess one thing that I could maybe try is to make the last message include the version it has been updating to:
```
fastlane-plugin-wpmreleasetoolkit have been updated to 1.3.7. Please check and commit the changes in your Gemfile.lock file, then restart your previous invocation of fastlane to use the new toolkit.
```
That version that `bundle update` would update to would depend on what you have in your `Pluginfile`, but most likely (since you're supposed to have `"~> x.y"` type of requirement there) it will be either the latest version if that latest version was `x.z` and not requiring a major version jump, or to an earlier version `x.z` if the latest version `u.v` was crossing the major version bump boundary (u>x).

Reply:
Thanks a lot for the explanation, it matches my expectations of what the code does. As you pointed out what is confusing to me is the messaging around it. Specifically, from your example:
```
Currently using release toolkit version 1.2.3
Checking for updates now...
There is a new version 2.1.4 of the release toolkit!
The latest version introduces breaking changes!
Do you want to run bundle update now?
```

When I am answering this question, it feels like I'll be updating to `2.14` because that's the information I've been presented so far. Especially for experienced Ruby developers, this is probably not an issue, but I think it'd help to clarify the messaging a bit for people who have less experience with bundler. It'll also not require the user to know whether the dependency is setup as `~> x.y` or not. How about something like this:

```
There is a new version 2.1.4 of the release toolkit!
The latest version introduces breaking changes!
Running bundle update will bump the version to 1.3.7, would you like to do it now?
```

I assume that this will make it obvious that a major bump needs to be manually handled, but if we want to include that in the messaging, we can. It's hard to balance between being too verbose and not providing enough info.

I'll go ahead and approve this and will leave the decision of how best to move forward to you. Thanks again!

Reply:
Yep and tbh that's also the message I wanted to print, but unfortunately there's no easy way to know what bundler will update the version to (1.3.7) before actually running it.

I mean there surely are ways to figure it out but that will not be as trivial as reading a property that would give us that value, but instead require us to run `bundle outdated` and have bundler fetch all the available versions from the server, then solve the dependency graph according to what is set in the Gemfile and the Pluginfile to determine which version is compatible with the constraint set there, to finally get that value. All the things (contacting server, resolving dev graph…) that `bundle update` will then redo when we finally invoke it to do the _actual_ update.

I'll see if we can get a better guess using the Updater class build in fastlane's codebase to get a better guess at which version it might update to — eg by assuming the Pluginfile will always be "~> x.y" and by assuming the dependency graph will never have another gem requirement that is a dependency of the toolkit blocking the the update, etc, so will still be a guess — or if not, if I can improve the message to make it clearer that bundle update might not update to the very latest if breaking changes are involved.

Reply:
Turns out even `bundle outdated` doesn't tell you which version it is going to update to, but instead tells you the latest available version regardless of the dependency graph constraints too.

For example if I run `bundle outdated` on WPiOS now, I get amongst other things:
```
$ bundle outdated --parseable

activesupport (newest 6.1.3.2, installed 5.2.6)
bigdecimal (newest 3.0.2, installed 1.4.4)
fastimage (newest 2.2.4, installed 2.2.3)
…
```
And you can see that `bigdecimal` is listed to have a newest version of 3.0.2. But if I run `bundle update bigdecimal`, it doesn't update it at all – because in our context, `bigdecimal` comes into our bundle by being a dependency on the `fastlane-plugin-wpmreleasetoolkit`, and that dependency is constrained to be `~> 1.4` in the toolkit's `gemspec`, so `bundle update` won't update to anything greated than `1.x` with `x>4`, and indeed `1.4.4` is the latest of the `1.x` version for that gem:
```
$ bundle update bigdecimal
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies...
…
Bundler attempted to update bigdecimal but its version stayed the same
Bundle updated!
```

---

So even `bundle outdated` doesn't know which version it is gonna update to until you actually run `bundle update` and it does the whole "resolving dependency graph" dance (and resolving a dependency graph is not always trivial indeed!), and there's apparently no easy way to determine that version beforehand.